### PR TITLE
`Programming exercises`: Fix an issue with participant scores when resetting an exercise or an exam

### PR DESCRIPTION
**Pull request by stefanwaldhauser**
#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [X] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [X] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [X] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [X] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [X] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [X] I implemented the changes with a good performance and prevented too many database calls.
- [X] I documented the Java code using JavaDoc style.

### Motivation and Context

When resetting programming or quiz exercises we sometimes had the following bug: 
```
Aug 02 22:02:29 artemis-production-node11.artemis.in.tum.de java[2073370]: 2022-08-02 22:02:29.151 ERROR 2073370 --- [nio-8080-exec-6] o.h.e.jdbc.batch.internal.BatchingBatch  : HHH000315: Exception executing batch [java.sql.BatchUpdateException: Lock wait timeout exceeded; try restarting transaction], SQL: update participant_score set exercise_id=?, last_points=?, last_rated_points=?, last_rated_result_id=?, last_rated_score=?, last_result_id=?, last_score=?, user_id=? where id=?
Aug 02 22:02:29 artemis-production-node11.artemis.in.tum.de java[2073370]: 2022-08-02 22:02:29.153  WARN 2073370 --- [nio-8080-exec-6] o.h.engine.jdbc.spi.SqlExceptionHelper   : SQL Error: 1205, SQLState: 40001
Aug 02 22:02:29 artemis-production-node11.artemis.in.tum.de java[2073370]: 2022-08-02 22:02:29.154 ERROR 2073370 --- [nio-8080-exec-6] o.h.engine.jdbc.spi.SqlExceptionHelper   : Lock wait timeout exceeded; try restarting transaction
Aug 02 22:02:29 artemis-production-node11.artemis.in.tum.de java[2073370]: 2022-08-02 22:02:29.183  WARN 2073370 --- [nio-8080-exec-6] o.z.problem.spring.common.AdviceTraits   : Conflict: could not execute batch; SQL [update participant_score set exercise_id=?, last_points=?, last_rated_points=?, last_rated_result_id=?, last_rated_score=?, last_result_id=?, last_score=?, user_id=? where id=?]; nested exception is org.hibernate.PessimisticLockException: could not execute batch
```

The exact reason for this is unclear, but it seems connected to the update of associated participant scores when resetting an exercise. 

**To hopefully prevent this @krusche had the idea to delete all participant scores belonging to an exercise before we remove the participations/results and thuse causing the update of participant score**

This might prevent the bug from happening again.

### Description

Removal of all connected participant scores when resetting an exercise

### Steps for Testing

Test resetting of quiz and programming exercises in particular if that stil works

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### Test Coverage
Extended Integration Test for Reset to confirm that participant scores are deleted.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/ExerciseDeletionService.java [hunk 1]
```diff
@@ -221,6 +221,9 @@ public void reset(Exercise exercise) {
      * @param exercise for which the plagiarism results and participations should be deleted
      */
     public void deletePlagiarismResultsAndParticipations(Exercise exercise) {
+        // delete all participant scores to avoid issues when deleting results later on
+        participantScoreRepository.deleteAllByExerciseIdTransactional(exercise.getId());
```
Comment (canberkanar) [inline src/main/java/de/tum/in/www1/artemis/service/ExerciseDeletionService.java:225]: Please verify with @krusche that the use of transactional is appropriate here.

**Review by canberkanar — APPROVED** [review]
Added a minor note. Other than that, the code looks good 👍🏻

File: src/main/java/de/tum/in/www1/artemis/service/ExerciseDeletionService.java (hunk 1 above)
Comment (Hialus) [reply src/main/java/de/tum/in/www1/artemis/service/ExerciseDeletionService.java:225]: This was pushed by @krusche, and considering that this is a larger delete operation Transaction should be fine here.

**Review by Hialus — APPROVED** [review]
Code changes look good to me.

File: src/main/java/de/tum/in/www1/artemis/service/ExerciseDeletionService.java (hunk 1 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/ExerciseDeletionService.java:225]: Adding to that: we did not add a new transactional here, but we simply use an existing one.

But most probably the nested transaction annotations led to problems in this particular case

**Review by krusche — APPROVED** [review]
I found the issue. When there are multiple results for the same participation (which is very common for programming and quiz exercises), deleting results within the reset functionality does not work due to a deadlock of 2 simultaneous running transactions.

I verified this with a minimal example on TS0 with the current state of `develop` which ran into the exception described in the PR.

When executing the same `reset exam` functionality with the changes in this PR it works correctly, so we can merge this.

As a follow-up: we need to verify this root cause for quizzes as well (e.g. on result in the live mode and one in the practice mode) and then find a similar solution here as well, when re-evaluating the quiz with changes for results we could simply delete all participant scores, because when saving the results, they will be created again anyway.

**Pull request merged**